src/azure-cli/azure/cli/command_modules/ams/operations/mru.py

```diff
@@ -84,24 +90,34 @@ def _get_v2_api_endpoint(self, cli_ctx, resource_group_name, account_name):
         api_endpoint = next((x for x in api_endpoints if x.get('majorVersion') == '2'), api_endpoints[0])
         if not api_endpoint:
             raise CLIError('v2 Media API endpoint was not found.')
-
         return api_endpoint
 
     def _get_v2_access_token(self, cli_ctx):
         from adal import AuthenticationContext
-
-        refresh_token_obj = self.profile.get_refresh_token()
-
-        refresh_token = refresh_token_obj[1]
-        tenant = refresh_token_obj[3]
         # pylint: disable=protected-access
-        client_id = self.profile.get_login_credentials()[0]._token_retriever()[2]['_clientId']
 
+        raw_token_obj = self.profile.get_raw_token()
+        raw_token = raw_token_obj[0][2]
+        clientId = raw_token['_clientId']
+        tenant = raw_token_obj[2]
         authority = '{}/{}'.format(cli_ctx.cloud.endpoints.active_directory, tenant)
         context = AuthenticationContext(authority)
-        return context.acquire_token_with_refresh_token(refresh_token,
-                                                        client_id,
-                                                        self.v2_media_api_resource).get('accessToken')
+        account = self.profile.get_subscription()
+        user_type = account[_USER_ENTITY][_USER_TYPE]
+
+        if user_type == _USER:
+            refresh_token = raw_token[_REFRESH_TOKEN]
+            return context.acquire_token_with_refresh_token(refresh_token,
+                                                            clientId,
+                                                            self.v2_media_api_resource).get(_ACCESS_TOKEN)
+        if user_type == _SERVICE_PRINCIPAL:
+            sp_authinfo = self.profile.get_sp_auth_info()
+            acq_token = context.acquire_token_with_client_credentials(self.v2_media_api_resource,
+                                                                      sp_authinfo['clientId'],
+                                                                      sp_authinfo['clientSecret'])
+            return acq_token['accessToken']
```

Comment:
I believe the logics here can be replaced with one line call of `get_row_token(resource=self.v2_media_api_resource)`

Reply:
This statement works for both user logins and service principal: self.profile.get_raw_token(resource=self.v2_media_api_resource)[0][2].get(_ACCESS_TOKEN)
Do you recommend **NOT** using _ACCESS_TOKEN from _profile.py?

Reply:
You should not take a dependency on anything that is private (leading _) as it explicitly makes no guarantee of stability.

Reply:
@tjprescott Thanks Travis. I'll use a local string.